### PR TITLE
Added Settings UI

### LINKED ISSUE #38: Settings Screen to allow configuring the Plugin without code
**Is your enhancement related to a problem? Please describe.**
Yes. The plugin can't be extended to custom post types without code.

**Describe the solution you'd like**
A settings page which allows users to configure the plugin would make it easier to install & adapt the plugin. MVP would be a settings section which allows uses to select the list of Post Types that should use Convert to blocks.

**Designs**
A multi-checkbox should suffice.

**Describe alternatives you've considered**
NA

**Additional context**
A [Filter hook was needed to tell the plugin to load](https://wordpress.org/support/topic/plugin-doesnt-work-at-all-9/) on custom post types.

### LINKED ISSUE #54: Add settings UI to opt CPTs into convert-to-blocks support
**Describe the bug**

When enabled, my custom post type, even when set to supports:editor is loading with the Classic Editor, not Gutenberg

**Steps to Reproduce**

1. Create a custom post type using CPTUI - set it to support editor
2. Create a new post 
3. Classic editor is displayed

**Expected behavior**

Gutenberg editor should be displayed

**Screenshots**

https://drive.google.com/file/d/18g7ZA3PO_7z0SYCS988aDoY3gX1bGHTR/view?usp=sharing

https://drive.google.com/file/d/18guVcMb80axpMMwVhFf1Dno69k2G9YIQ/view?usp=sharing

https://drive.google.com/file/d/18hc4ye0AJTdkV9jFrJTyrVGPhomfXxRs/view?usp=sharing

**Environment information**

WordPress 5.7

### PR DESCRIPTION
**Pull request by akshitsethi**
Closes #38, #54

### Description of the Change
This PR adds UI for Plugin settings to configure the `post_types` supported by the plugin. Currently, it is being done with the help of a filter hook. With this change, users will be able to visually select the `post_types` they wish to be configured with the plugin.

Also, missing [yoast/phpunit-polyfills](https://packagist.org/packages/yoast/phpunit-polyfills) package has been added to enable running of unit tests.

**To-do:**

- [x] Polishing PR
- [x] Configure filter to use `post_types` value from DB to configure the plugin

### Possible Drawbacks

None that I can think of.

### Verification Process

To check the working of the Settings panel, follow the steps below:

- Click "Convert to Blocks" under the "Settings" menu.
- On the **Settings** page, de-select one of the default post types, i.e. either page or post.
- Navigate to the de-selected item's menu (Page, Post or CPT) and verify that it is no longer supported by the plugin.

### Checklist:

- [x] I have read the [**CONTRIBUTING**](/CONTRIBUTING.md) document.
- [x] My code follows the code style of this project.
- [x] All new and existing tests passed.

### Changelog Entry

> Added - Settings UI for managing supported **post_types**

### CODE REVIEW

File: includes/ConvertToBlocks/Settings.php [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+	 */
+	public function register() {
+		// Configure variables and get post types.
+		$this->init();
+
+		add_action( 'admin_menu', [ $this, 'add_menu' ] );
+		add_action( 'admin_init', [ $this, 'register_section' ], 10 );
+		add_action( 'admin_init', [ $this, 'register_fields' ], 20 );
+
+		add_filter( 'post_type_supports_convert_to_blocks', [ $this, 'supported_post_types' ], PHP_INT_MAX, 2 );
```
Comment (dinhtungdu) [inline includes/ConvertToBlocks/Settings.php]: Using `PHP_INT_MAX` makes it very hard for plugins and/or themes to use this filter. I think even using the default priority is enough.

**Review by dinhtungdu — COMMENTED** [review]
@akshitsethi thanks so much for the PR! Testing this PR I found some issues:
- Activating the plugin for the first time, the Post and Page are unchecked in the setting page, while they're enabled by default. Pages and posts are still converted to blocks automatically. This can make users confused.
- This doesn't fully resolve #54 because the custom post types with `editor` support are still forced to use the classic editor.

File: includes/ConvertToBlocks/Settings.php (hunk 1 above)
Comment (akshitsethi) [reply includes/ConvertToBlocks/Settings.php]: @dinhtungdu I totally agree. I was myself having second thoughts about this. I'll correct this.

**Comment by akshitsethi** [comment]
@dinhtungdu
> Activating the plugin for the first time, the Post and Page are unchecked in the setting page, while they're enabled by default. Pages and posts are still converted to blocks automatically. This can make users confused.

There are certain scenarios here which I would like a bit more clarity on:

#### Scenario 1

Post and pages are enabled by default. What if a user wants to disable the functionality on pages but keep it on for posts?
Should we factor this in the plugin or keep it enabled by default for both (without allowing to de-activate).

Another point to consider here is that if we do keep the settings enabled by default for the user, then on a blog without CPTs, we won't have any setting to configure.

#### Scenario 2

An existing user has configured the plugin using the filter but if we update plugin settings in DB enabling post and pages options by default, it might break the existing filter they have in place. One way to tackle this is to ensure that our filter runs first and then it can be overriden later.

Do you think updating DB option on plugin activation should be done?

**Comment by dinhtungdu** [comment]
> Post and pages are enabled by default. What if a user wants to disable the functionality on pages but keep it on for posts?
Should we factor this in the plugin or keep it enabled by default for both (without allowing to de-activate).

What I mean is we should respect the default setting. If users activate the plugin for the first time, Post and Page should be checked on the setting page (there is no data in the database at that point). If user have another CPTs, those post type should be unchecked by default.

> An existing user has configured the plugin using the filter but if we update plugin settings in DB enabling post and pages options by default, it might break the existing filter they have in place.

The custom filter added by users should have the highest priority and should disable the setting section. We also should display a message let users know that there is an active filter controlling the post type.

> Do you think updating DB option on plugin activation should be done?

Because of the above points, I don't think we need a DB updater here.

File: includes/ConvertToBlocks/Settings.php [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+
+	/**
+	 * Renders the `post_types` field.
+	 *
+	 * @return void
+	 */
+	public function field_post_types() {
+		$post_types  = get_option(
+			sprintf( '%s_post_types', CONVERT_TO_BLOCKS_SLUG ),
+			apply_filters( 'convert_to_blocks_default_post_types', CONVERT_TO_BLOCKS_DEFAULT_POST_TYPES )
```
Comment (dinhtungdu) [inline includes/ConvertToBlocks/Settings.php]: We should avoid using the same filter in different places when possible.

For this case, we can define the `$cointainer` property for this class, and get the default post types using `$this->container->get_default_post_types(), see [this](https://github.com/10up/convert-to-blocks/blob/d33e248866170f2d0372b72e83d8340b1eb13d11/includes/ConvertToBlocks/Plugin.php#L148-L157) for more details.

By doing so, we don't need the `CONVERT_TO_BLOCKS_DEFAULT_POST_TYPES` constant anymore.

File: includes/ConvertToBlocks/Settings.php [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+			[ 'page' => CONVERT_TO_BLOCKS_SLUG ],
+			admin_url( 'options-general.php' )
+		);
+		?>
+		<div class="notice notice-success is-dismissible">
+			<p>
+				<?php
+				printf(
+					/* translators: %1$s: link to switch to settings panel, %2$s: closing anchor tag */
+					esc_html__( 'A filter hook (post_type_supports_convert_to_blocks) is active which can lead to undesirable outcome. Kindly remove it and configure settings via the %1$sSettings Panel%2$s.', 'convert-to-blocks' ),
```
Comment (dinhtungdu) [inline includes/ConvertToBlocks/Settings.php]: > `Kindly remove it and configure settings via the %1$sSettings Panel%2$s.`

~The notice is only displayed on the Convert to Blocks setting page, so we don't need this IMO.~

Edit: the notice is displayed for every admin page. I think we should limit it to the post/page list, plugin page, and the Convert to Blocks setting pages.

Comment (dinhtungdu) [reply includes/ConvertToBlocks/Settings.php]: Also, a filter can be used for advanced conditions like preventing conversion for posts belonging to a category. So IMHO, we just need to notify users that one or more filters to `post_type_supports_convert_to_blocks` are being used, without any call to action.

**Comment by jeffpaul** [comment]
@akshitsethi are you able to work on the feedback items here or would it be best to grab another person to help wrap up this PR?

**Comment by Sidsector9** [comment]
@jeffpaul, @akshitsethi said this needs a little more work and he's happy to have someone from the OSP team to takeover.

**Comment by cadic** [comment]
@Sidsector9 @jeffpaul will work on this during my OSP week.

**Comment by jeffpaul** [comment]
@cadic did you manage to make progress on the updates here locally and can push up to the PR?

**Comment by jeffpaul** [comment]
@Sidsector9 @cadic checking to see if either of you are able to help get this updated and off to code review in January perhaps?

**Review by Sidsector9 — DISMISSED** [review]
LGTM 👍 

I added few fixes for workflows for them to pass, downgraded some NPM dependencies as `npm run build` broke.

@jeffpaul should we bump the minimum PHP version of this plugin from 7.0 to 7.4?

**Comment by jeffpaul** [comment]
@Sidsector9 yes, fine to bump PHP min to 7.4

**Review by Sidsector9 — APPROVED** [review]

**Pull request merged**